runtime/runtime_impl_jetty9/src/main/java/com/google/apphosting/runtime/jetty9/AppVersionHandlerFactory.java

```diff
@@ -199,24 +201,34 @@ private Handler doCreateHandler(AppVersion appVersion) throws ServletException {
       SessionManagerHandler unused = SessionManagerHandler.create(builder.build());
       // Pass the AppVersion on to any of our servlets (e.g. ResourceFileServlet).
       context.setAttribute(AppEngineConstants.APP_VERSION_CONTEXT_ATTR, appVersion);
-
-      context.start();
-      // Check to see if servlet filter initialization failed.
-      Throwable unavailableCause = context.getUnavailableException();
-      if (unavailableCause != null) {
-        if (unavailableCause instanceof ServletException) {
-          throw (ServletException) unavailableCause;
-        } else {
-          UnavailableException unavailableException =
-              new UnavailableException("Initialization failed.");
-          unavailableException.initCause(unavailableCause);
-          throw unavailableException;
-        }
+      if (Boolean.getBoolean(AppEngineConstants.HTTP_CONNECTOR_MODE)) {
+        context.addEventListener(
+            new ContextHandler.ContextScopeListener() {
+              @Override
+              public void enterScope(
+                  ContextHandler.Context context,
+                  org.eclipse.jetty.server.Request request,
+                  Object reason) {
+                if (request != null) {
+                  ApiProxy.Environment environment =
+                      (ApiProxy.Environment) request.getAttribute(AppEngineConstants.ENVIRONMENT_ATTR);
+                  if (environment != null) {
+                    ApiProxy.setEnvironmentForCurrentThread(environment);
+                  }
+                }
+              }
+
+              @Override
+              public void exitScope(
+                  ContextHandler.Context context, org.eclipse.jetty.server.Request request) {
+                if (request != null) {
```

Comment:
Should this also check that there is a ENVIRONMENT_ATTR set?

Reply:
We always set the attribute in the `JettyHttpHandler` so it will always be set.